tools/diagram-sparql

```diff
@@ -162,22 +189,99 @@ while (1) {
 		# print $all; exit 0;
 		next;
 		}
-	if ($terminator eq ".") {
+	if ($v =~ m/\A\(([^\s\)].*)/) {
+		# Non-empty list, but "(" needs a space after it.  
+		# Add one and try again.
+		$v = $1;
+		$all = "$s $p ( $v $terminator $all";
+		warn "Open paren needs a space after it: ($v\n" if $debug;
+		next;
+		}
+	elsif ($terminator eq ".") {
 		die "$0: ERROR: Parse got multi-line value: \n[[\n$v\n]]\n"
 			if $v =~ m/\n/s;
 		# Turn whitespace into underscore:
 		$v =~ s/\s/_/g;
 		push(@triples, [$s, $p, $v]);
+		warn "Assertion: $s $p $v\n" if $debug;
 		}
 	# Expand ";" and "," operators to full triples:
 	elsif ($terminator eq ";") {
 		# Push the expanded statement back on the front to be reparsed:
 		$all = "$s $p $v .\n$s $all";
+		warn "Expanding ';' after $v\n" if $debug;
 		next;
 		}
 	elsif ($terminator eq ",") {
 		# Push the expanded statement back on the front to be reparsed:
 		$all = "$s $p $v .\n$s $p $all";
+		warn "Expanding ',' after $v\n" if $debug;
+		next;
+		}
+	elsif ($v eq "(" && $terminator =~ m/\A\)/) {
+		# Empty list, but with a space between parens: ( ).
+		# Remove the space and try again.
+		$terminator = $';
+		$all = "$s $p () $terminator $all";
+		warn "Empty list with space: ( )\n" if $debug;
+		next;
+		}
+	elsif ($v eq "(") {
+		# Non-empty list.  Flatten into separate statements
+		# and re-parse.
+		warn "Non-empty list.  Time to flatten...\n" if $debug;
+		my $listAll = "$terminator $all";
+		# Collect all list elements.
+		# See perl Regex::Common for $RE usage and
+		# http://stackoverflow.com/questions/7974093/matching-balanced-parenthesis-in-perl-regex
+		my $bp = $RE{balanced}{-parens=>'()'};
+		while (1) {
```

Comment:
Not crazy about while (1) and using last to break out as a style though the use here is fairly clean since there is just one exit path.